In [1]:
import numpy as np

In [2]:
np.__version__

'1.26.4'

We will use the same array as in the previous exercises, which represents the performance of 5 athletes in 4 laps of a running track, on 3 different days.

In [3]:
rng = np.random.default_rng(seed=1)

In [4]:
array = rng.normal(loc=50, scale=3, size=60).reshape(3, 5, 4).round(3)

In [5]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [6]:
array.shape

(3, 5, 4)

👆 Notice from the `shape` that there are 3 days, 5 athletes and 4 laps.

#### `max` / `argmax`

Remember that, as the values ​​represent the time, in seconds, that each athlete took to complete a lap, the higher this value, the worse the athlete's performance.

Use `max` to find the worst daily performance (axis 0).

In [7]:
array.max(axis=0)

array([[51.037, 52.465, 53.882, 53.02 ],
       [52.716, 51.339, 50.107, 51.743],
       [51.781, 52.674, 56.354, 51.64 ],
       [52.195, 56.128, 52.637, 51.989],
       [52.743, 49.94 , 50.502, 50.327]])

👆 The result is an array with 5 rows (corresponding to the 5 athletes) and 4 columns (corresponding to the 4 laps). Each value is the maximum time over the 3 days for that athlete and that lap.

Repeat the exercise with `argmax`.

In [8]:
array.argmax(axis=0)

array([[0, 0, 1, 1],
       [0, 0, 2, 0],
       [2, 2, 1, 0],
       [2, 1, 2, 1],
       [2, 2, 1, 1]])

👆 This indicates that, for the first athlete and the first lap, (first row and first column of the visualization) the worst performance was on day 0; for the last athlete and the last lap, the worst performance was on day 1.

Now use `max` and `argmax` to find the worst performance per athlete (axis 1), on each day and on each lap.

In [9]:
array.max(axis=1)

array([[52.716, 52.465, 50.991, 51.797],
       [50.641, 56.128, 56.354, 53.02 ],
       [52.743, 52.674, 52.637, 49.058]])

👆 This is the worst performance among all athletes, in the 3 days (rows) and in the 4 laps (columns).

In [10]:
array.argmax(axis=1)

array([[1, 0, 0, 3],
       [2, 3, 2, 0],
       [4, 2, 3, 4]])

👆 On the first day (first row), in the 4 laps (columns), the worst athletes were those with index 1, 0, 0 and 3, respectively.

Also find the worst overall performance.

In [11]:
array.max()

56.354

#### `min`/`argmin`

Repeat the exercise, now with the `mix`/`argmix` methods, to find the best performances.

Interpret the results in a similar way to that done for the worst performers.

In [13]:
array.min(axis=0)

array([[46.318, 47.95 , 49.784, 46.091],
       [41.867, 44.333, 48.389, 48.481],
       [50.641, 50.652, 50.085, 46.664],
       [47.791, 48.496, 48.554, 46.785],
       [48.458, 45.056, 46.254, 49.058]])

In [14]:
array.argmin(axis=0)

array([[2, 2, 2, 0],
       [1, 1, 0, 2],
       [1, 1, 0, 1],
       [0, 2, 0, 2],
       [1, 1, 2, 2]])

In [15]:
array.min(axis=1)

array([[47.791, 49.123, 47.654, 46.091],
       [41.867, 44.333, 49.476, 46.664],
       [46.318, 47.95 , 46.254, 46.785]])

In [16]:
array.argmin(axis=1)

array([[3, 4, 4, 0],
       [1, 1, 1, 2],
       [0, 0, 4, 3]])

In [17]:
array.min()

41.867

#### `ptp`

Find the difference between the worst and best performance, on each day (axis 0).

In [18]:
array.ptp(axis=0)

array([[ 4.719,  4.515,  4.098,  6.929],
       [10.849,  7.006,  1.718,  3.262],
       [ 1.14 ,  2.022,  6.269,  4.976],
       [ 4.404,  7.632,  4.083,  5.204],
       [ 4.285,  4.884,  4.248,  1.269]])

👆 In the case of athlete 0 (first row), his performance fluctuated 4.719 seconds in the first lap, 4.515 seconds in the second lap, and so on.

Find the difference between the worst and best performance, in each lap (axis 2).

In [19]:
array.ptp(axis=2)

array([[6.374, 4.327, 1.555, 4.006, 2.465],
       [4.709, 7.609, 9.69 , 7.261, 5.446],
       [3.466, 1.805, 5.129, 5.852, 6.489]])

👆 In the case of lap indexed 1 (second column), the difference in performance between the athletes on the first day was 4.327 seconds, 7.609 seconds on the second day and 1.805 seconds on the third.

#### `clip`

"Clip" this array so that no value is less than 43 or greater than 55. Display the array to view the changes.

In [20]:
array_clip = array.clip(min=43, max=55)

In [21]:
array_clip

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [43.   , 44.333, 49.476, 48.733],
        [50.641, 50.652, 55.   , 46.664],
        [48.867, 55.   , 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Confirm with `min` and `max` that the change was made as desired.

In [22]:
array_clip.min()

43.0

In [23]:
array_clip.max()

55.0

#### `round`

Use `round` to round values ​​to 2 decimal places.

In [24]:
array.round(decimals=2)

array([[[51.04, 52.46, 50.99, 46.09],
        [52.72, 51.34, 48.39, 51.74],
        [51.09, 50.88, 50.08, 51.64],
        [47.79, 49.51, 48.55, 51.8 ],
        [50.12, 49.12, 47.65, 49.23]],

       [[50.02, 49.17, 53.88, 53.02],
        [41.87, 44.33, 49.48, 48.73],
        [50.64, 50.65, 56.35, 46.66],
        [48.87, 56.13, 51.94, 51.99],
        [48.46, 45.06, 50.5 , 50.33]],

       [[46.32, 47.95, 49.78, 47.17],
        [49.7 , 50.29, 50.11, 48.48],
        [51.78, 52.67, 50.96, 47.54],
        [52.2 , 48.5 , 52.64, 46.78],
        [52.74, 49.94, 46.25, 49.06]]])

Also check rounding to 1 decimal place.

In [25]:
array.round(decimals=1)

array([[[51. , 52.5, 51. , 46.1],
        [52.7, 51.3, 48.4, 51.7],
        [51.1, 50.9, 50.1, 51.6],
        [47.8, 49.5, 48.6, 51.8],
        [50.1, 49.1, 47.7, 49.2]],

       [[50. , 49.2, 53.9, 53. ],
        [41.9, 44.3, 49.5, 48.7],
        [50.6, 50.7, 56.4, 46.7],
        [48.9, 56.1, 51.9, 52. ],
        [48.5, 45.1, 50.5, 50.3]],

       [[46.3, 48. , 49.8, 47.2],
        [49.7, 50.3, 50.1, 48.5],
        [51.8, 52.7, 51. , 47.5],
        [52.2, 48.5, 52.6, 46.8],
        [52.7, 49.9, 46.3, 49.1]]])

#### `sum`

Use `sum` to view the total time spent by each athlete in each race, over the 3 days. In other words, you must add the time on axis 2, corresponding to the laps.

In [26]:
array_sum_axis_2 = array.sum(axis=2)

In [27]:
array_sum_axis_2

array([[200.584, 204.187, 203.701, 197.653, 196.124],
       [206.099, 184.409, 204.311, 208.924, 194.343],
       [191.218, 198.579, 202.963, 200.113, 197.995]])

👆 The array represents the 3 days (3 rows) and the 5 athletes (5 columns).

Now, from this array, add up the performances of the 5 athletes over the 3 days (axis 0).

In [28]:
array_sum_axis_2_axis_0 = array_sum_axis_2.sum(axis=0)

In [29]:
array_sum_axis_2_axis_0

array([597.901, 587.175, 610.975, 606.69 , 588.462])

👆 The array represents the total sum of the performance of the 5 athletes.

#### `cumsum`

Use `cumsum` to find the accumulated time of each athlete, on each day, at the end of each lap (axis 2).

In [30]:
array.cumsum(axis=2)

array([[[ 51.037, 103.502, 154.493, 200.584],
        [ 52.716, 104.055, 152.444, 204.187],
        [ 51.094, 101.976, 152.061, 203.701],
        [ 47.791,  97.302, 145.856, 197.653],
        [ 50.119,  99.242, 146.896, 196.124]],

       [[ 50.024,  99.197, 153.079, 206.099],
        [ 41.867,  86.2  , 135.676, 184.409],
        [ 50.641, 101.293, 157.647, 204.311],
        [ 48.867, 104.995, 156.935, 208.924],
        [ 48.458,  93.514, 144.016, 194.343]],

       [[ 46.318,  94.268, 144.052, 191.218],
        [ 49.705,  99.991, 150.098, 198.579],
        [ 51.781, 104.455, 155.418, 202.963],
        [ 52.195, 100.691, 153.328, 200.113],
        [ 52.743, 102.683, 148.937, 197.995]]])

👆 On the second day, the third athlete (row 2) accumulated times of 50.641, 101.293, 157.647 and 204.311.

#### `mean` / `var` / `std`

Find the average performance per lap (axis 2) for each athlete, across the 3 days.

In [31]:
array.mean(axis=2)

array([[50.146  , 51.04675, 50.92525, 49.41325, 49.031  ],
       [51.52475, 46.10225, 51.07775, 52.231  , 48.58575],
       [47.8045 , 49.64475, 50.74075, 50.02825, 49.49875]])

👆 Average performance per lap over the 3 days (3 rows), for the 5 athletes (5 columns).

Also display the variance and standard deviation of the same data.

In [32]:
array.var(axis=2)

array([[ 5.832173  ,  2.60505119,  0.31180869,  2.26545619,  0.7817865 ],
       [ 3.89384469,  9.84265569, 11.92300419,  6.6615875 ,  4.79485519],
       [ 1.63923875,  0.49570519,  3.77044719,  6.09171069,  5.36063069]])

In [33]:
array.std(axis=2)

array([[2.41498923, 1.6140171 , 0.55839832, 1.50514324, 0.88418691],
       [1.97328272, 3.1373007 , 3.45297034, 2.58100513, 2.18971578],
       [1.28032759, 0.70406334, 1.94176394, 2.46813911, 2.31530358]])

Now display these 3 statistical measures, aggregated per athlete (axis 1).

In [34]:
array.mean(axis=1)

array([[50.5514, 50.664 , 49.1346, 50.0998],
       [47.9714, 49.0684, 52.4308, 50.1466],
       [50.5484, 49.8692, 49.949 , 47.807 ]])

👆 Average performance per athlete over the 3 days (3 rows), over the 4 laps (4 columns).

In [35]:
array.var(axis=1)

array([[ 2.60449864,  1.490168  ,  1.48693464,  4.95666456],
       [ 9.92850184, 18.17600984,  6.03789536,  5.16203544],
       [ 5.53089424,  2.72292896,  4.3917508 ,  0.7086572 ]])

In [36]:
array.std(axis=1)

array([[1.61384592, 1.22072438, 1.21939929, 2.22635679],
       [3.15095253, 4.26333318, 2.45721293, 2.27201132],
       [2.35178533, 1.65012998, 2.09565045, 0.8418178 ]])

#### `all` / `any`

First, create a boolean mask (array with `True`/`False` values) to check which values ​​in `array` are greater than 50 seconds.

In [37]:
array_mask = array > 50

In [38]:
array_mask

array([[[ True,  True,  True, False],
        [ True,  True, False,  True],
        [ True,  True,  True,  True],
        [False, False, False,  True],
        [ True, False, False, False]],

       [[ True, False,  True,  True],
        [False, False, False, False],
        [ True,  True,  True, False],
        [False,  True,  True,  True],
        [False, False,  True,  True]],

       [[False, False, False, False],
        [False,  True,  True, False],
        [ True,  True,  True, False],
        [ True, False,  True, False],
        [ True, False, False, False]]])

Now, use `all` to check if any athlete took more than 50 seconds to complete a lap (axis 2), on any given day.

In [39]:
array_mask.all(axis=2)

array([[False, False,  True, False, False],
       [False, False, False, False, False],
       [False, False, False, False, False]])

👆 This indicates that, on the first day (first row), the athlete with index 2 (third column) took more than 50 seconds in all laps.

Use `any` to repeat the analysis, but considering the time spent on any lap.

In [40]:
array_mask.any(axis=2)

array([[ True,  True,  True,  True,  True],
       [ True, False,  True,  True,  True],
       [False,  True,  True,  True,  True]])

👆 On the first day, all athletes had at least one lap that took more than 50 seconds; on the second day, only the athlete indexed 1 did not perform worse than 50 seconds.